storey/flow.py

```diff
@@ -608,63 +611,98 @@ async def _handle_completed(self, event, response):
 
 
 class _Batching(Flow):
-    def __init__(self, max_events: Optional[int] = None, timeout_secs=None, **kwargs):
+    def __init__(
+        self,
+        max_events: Optional[int] = None,
+        timeout_secs: Optional[int] = None,
+        group_events: bool = False,
+        key: Optional[Union[str, callable]] = None,
+        **kwargs,
+    ):
         super().__init__(**kwargs)
 
-        self._max_events = max_events
-        self._event_count = 0
-        self._batch = []
-        self._batch_time = None
-        self._timeout_task = None
-
+        self._max_events: int = max_events
         self._timeout_secs = timeout_secs
+        self._group_events = group_events
+
         if self._timeout_secs is not None and self._timeout_secs <= 0:
-            raise ValueError('Batch timeout cannot be 0 or negative')
+            raise ValueError("Batch timeout cannot be 0 or negative")
+
+        self._event_count: Dict[Optional[str], int] = defaultdict(int)
+        self._batch: Dict[Optional[str], List[Any]] = defaultdict(list)
+        self._batch_time: Dict[Optional[str], datetime] = {}
+        self._timeout_task: Optional[Task] = None
+
+        self.key_extractor = None
+        if key and group_events:
+            if callable(key):
+                self.key_extractor = key
+            elif isinstance(key, str):
+                self.key_extractor = lambda element: element[key]
 
     async def _emit(self, batch, batch_time):
         raise NotImplementedError
 
     async def _terminate(self):
         pass
 
-    async def _sleep_and_emit(self):
-        await asyncio.sleep(self._timeout_secs)
-        await self._emit_batch()
+    async def _sleep_and_emit(self, key: Optional[str] = None):
+        time_delta = (
+            datetime.now(timezone.utc) - self._batch_time[key]
+        ).total_seconds()
+        if time_delta < self._timeout_secs:
+            await asyncio.sleep(self._timeout_secs - time_delta)
+        await self._emit_batch(key)
+        self._timeout_task = None
```

Comment:
rather than setting "yourself" to None you can just sleep till the next key to emit as long as there are keys in the dict

Reply:
Please elaborate